Added utility function get_referred_instances()

jesper-friis opened the pull request:
# Description
Added new utility function get_referred_instances() that recursively finds all instances that are directly or indirectly referred to by an instance (e.g. a collection or a reference property).

Needed by Musicode demo.

## Type of change
- [ ] Bug fix & code cleanup
- [x] New feature
- [ ] Documentation update
- [ ] Test update

## Checklist for the reviewer
This checklist should be used as a help for the reviewer.

- [ ] Is the change limited to one issue?
- [ ] Does this PR close the issue?
- [ ] Is the code easy to read and understand?
- [ ] Do all new feature have an accompanying new test?
- [ ] Has the documentation been updated as necessary?

sygout commented:
@jesper-friis the documention of the new functions should be improved. It is not very clear what is explore and the risk associated to recursive search.

There is also no test associated to the new functions (it would help understand as well). 

That merge could be accepted if there is an urgent need but then a new issue should be created to fix the previous points.

jesper-friis commented:
> @jesper-friis the documention of the new functions should be improved. It is not very clear what is explore and the risk associated to recursive search.
> 
> There is also no test associated to the new functions (it would help understand as well).
> 
> That merge could be accepted if there is an urgent need but then a new issue should be created to fix the previous points.

I added more documentation with an oral example. 

I agree that it would be good with a test, but I don't have time right now to add interesting ones with nested (and possible cyclic) references. Added issue #673 to address that.

sygout commented on bindings/python/utils.py:
```diff
@@ -516,7 +516,62 @@ def infer_dimensions(meta, values, strict=True):
+    referred to by `inst`.
+
+    This function follows instances referred to by collections and via
+    properties of type 'ref'.  Transaction parents are not included,
```
It is not clear for me what transaction parents mean in that context. Do we need that to clarify the goal of the function?

jesper-friis replied:
I added it to improve the documentation and also document what this function doesn't do, since transactions also refer to other instances (in addition to collections, ref types and metadata). I think it is ok to keep - the reader can just ignore the sentence is he/she doesn't know what a transaction is.

jesper-friis commented on bindings/python/utils.py:
```diff
@@ -516,7 +516,62 @@ def infer_dimensions(meta, values, strict=True):
+    -------
+    If you have a collection 'coll' with three instances 'inst1',
+    'inst2' and 'inst3' and 'inst2' has a 'ref' property, which is
+    an array of 'inst4' and 'inst5', then
```
done

jesper-friis commented on bindings/python/utils.py:
```diff
@@ -516,7 +516,62 @@ def infer_dimensions(meta, values, strict=True):
+
+    would return the set `{coll, inst1, inst2, inst3, inst4, inst5}`.
+
+    It same set would be returned even if one of the instances would
```
done
